### PR TITLE
feat(auth-validator): [Auth/PM-22975] Client Version Validator

### PR DESCRIPTION
**Pull request by Patrick-Pimentel-Bitwarden**
# **NEEDS TESTING ON FEATURE BRANCH**

## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-22975

## 📔 Objective

The objective of this ticket is to create a client version validator that checks the header data on a token request and prevent users who have migrated to v2 from logging in with older client versions.

1. Extracted out KM utilities for checking isV2Encryption from RotateUserAccountKeysCommand
2. Added some clarity and comments in BaseRequestValidator
3. Installed the validation for client version in the appropriate place to prevent account enumeration

## 📸 Screenshots

https://github.com/user-attachments/assets/cf4bc61a-3ec8-4d28-a1c1-19690b755f6a

https://github.com/user-attachments/assets/36f51ae3-5bb0-4b65-9df3-595048b1cfbf

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs [hunk 1]
```diff
@@ -119,6 +123,13 @@ await BuildSuccessResultAsync(validatorContext.User, context, validatorContext.D
                 return;
             }
 
+            // 1.5 Now check the version number of the client. Do this after ValidateContextAsync so that
+            // we prevent account enumeration. If we were to do this before ValidateContextAsync, then attackers
+            // could use a known invalid client version and make a request for a user (before we know if they have
+            // demonstrated ownership of the account via correct credentials) and identify if they exist by getting
+            // an error response back from the validator saying the user is not compatible with the client.
```
Comment (Patrick-Pimentel-Bitwarden) [inline src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs]: 💭 Could someone double check my explanation here?

File: src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs [hunk 2]
```diff
@@ -119,6 +123,13 @@ await BuildSuccessResultAsync(validatorContext.User, context, validatorContext.D
...
+            await ValidateClientVersionAsync(context, validatorContext);
```
[the remark this answers is not in the document]
Comment (Patrick-Pimentel-Bitwarden) [reply src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs]: fixed with [91af02b](https://github.com/bitwarden/server/pull/6588/commits/91af02b9d28381430b1d617b94498a9e7d8d3ff1)

File: src/Core/KeyManagement/Utilities/EncryptionParsing.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+public static class EncryptionParsing
+{
+    /// <summary>
+    /// Helper method to convert an encryption type string to an enum value.
+    /// </summary>
+    public static EncryptionType GetEncryptionType(string encString)
+    {
+        var parts = encString.Split('.');
+        if (parts.Length == 1)
```
[the remark this answers is not in the document]
Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Utilities/EncryptionParsing.cs:15]: This is code copied from KM, I'm hesitant to make more changes to their work.

File: test/Identity.IntegrationTest/Login/ClientVersionGateTests.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    public async Task TokenEndpoint_GrantTypePassword_V2User_OnOldClientVersion_Blocked(RegisterFinishRequestModel requestModel)
+    {
+        var localFactory = new IdentityApplicationFactory();
+        var server = localFactory.Server;
+        var user = await localFactory.RegisterNewIdentityFactoryUserAsync(requestModel);
+
+        // Make user V2: set private key to COSE and add signature key pair
+        var db = localFactory.GetDatabaseContext();
+        var efUser = await db.Users.FirstAsync(u => u.Email == user.Email);
+        efUser.PrivateKey = "7.cose";
```
[the remark this answers is not in the document]
Comment (Patrick-Pimentel-Bitwarden) [reply test/Identity.IntegrationTest/Login/ClientVersionGateTests.cs]: Read and not relevant

**Comment by Patrick-Pimentel-Bitwarden** [comment]
I am seeing 100 failures in tests, double checking what is going on, consider holding off on approving.

File: src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+using Bit.Core.KeyManagement.Queries.Interfaces;
+using Bit.Core.KeyManagement.Repositories;
+using Bit.Core.KeyManagement.Utilities;
+
+namespace Bit.Core.KeyManagement.Queries;
+
+public class IsV2EncryptionUserQuery(IUserSignatureKeyPairRepository userSignatureKeyPairRepository)
+    : IIsV2EncryptionUserQuery
+{
+    public async Task<bool> Run(User user)
```
Comment (quexten) [inline src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: Nice, this query makes sense and is well re-usable.

File: src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        ArgumentNullException.ThrowIfNull(user);
+
+        var hasSignatureKeyPair = await userSignatureKeyPairRepository.GetByUserIdAsync(user.Id) != null;
+        var isPrivateKeyEncryptionV2 =
+            !string.IsNullOrWhiteSpace(user.PrivateKey) &&
+            EncryptionParsing.GetEncryptionType(user.PrivateKey) == EncryptionType.XChaCha20Poly1305_B64;
+
+        return hasSignatureKeyPair switch
```
Comment (quexten) [inline src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: To be honest, this is nit-picky, but this would be more readable as an if/else. If this were rust matching this would be different but it took me a second to get this.
i.e 
```
if (isPrivateKeyEncryptionV2 && hasSignatureKeyPair) {

} else if (!isPrivateKeyEncryptionV2 && !hasSignatureKeyPair) {

} else {
  throw
}
```

File: src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+public class IsV2EncryptionUserQuery(IUserSignatureKeyPairRepository userSignatureKeyPairRepository)
+    : IIsV2EncryptionUserQuery
+{
+    public async Task<bool> Run(User user)
+    {
+        ArgumentNullException.ThrowIfNull(user);
+
+        var hasSignatureKeyPair = await userSignatureKeyPairRepository.GetByUserIdAsync(user.Id) != null;
+        var isPrivateKeyEncryptionV2 =
```
Comment (quexten) [inline src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: Note: This logic is fine, but a simpler way would be to check the user's security version property. If it is >=2 it is a encryption v2 user, else (null) a V1 user. That way you could even just implement a function on the user class "IsV2Encryption()" and call user.IsV2Encryption()

File: src/Core/KeyManagement/Utilities/EncryptionParsing.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (quexten) [reply src/Core/KeyManagement/Utilities/EncryptionParsing.cs]: Yeah, this code was inherited when KM formed. But looks reasonable.

File: src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs (hunk 1 above)
Comment (quexten) [reply src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs]: Is the validatorContext.User value even set before validating? (I'm not sure)

Explanation makes sense though and is very helpful.

File: src/Core/KeyManagement/Utilities/EncryptionParsing.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (quexten) [reply src/Core/KeyManagement/Utilities/EncryptionParsing.cs:15]: Agree, best to ignore claude here, you are just extracting existing code.

File: src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs [hunk 8]
```diff
@@ -0,0 +1,31 @@
+﻿using Bit.Core.Entities;
+using Bit.Core.Enums;
+using Bit.Core.KeyManagement.Queries.Interfaces;
+using Bit.Core.KeyManagement.Repositories;
+using Bit.Core.KeyManagement.Utilities;
+
+namespace Bit.Core.KeyManagement.Queries;
+
+public class IsV2EncryptionUserQuery(IUserSignatureKeyPairRepository userSignatureKeyPairRepository)
```
Comment (quexten) [inline src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: Ok, so having thought about this a bit, I think we should remove this query. We can do either `user.GetSecurityVersion() >= 2` or implement "IsV2User" on the user class which does the same.

Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: So I took this feedback and tweaked it a little. I introduced a narrower scoped check:

`user.IsSecurityVersionTwo()`

I felt this was more defensive because now we have breadcrumbed places where we are doing branching logic based on a V2 encryption requirement.

While I am happy to be using this simpler approach than the IsV2User query I want to recognize this is a deviation from the ticket's original tech breakdown and should this be discussed to confirm this wouldn't create any issues?

See the changes in [753670d](https://github.com/bitwarden/server/pull/6588/commits/753670d26fee9314f7502b70ccf8324ee480168e)

File: src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs (hunk 6 above)
Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: No longer needed with using the SecurityVersion column on the user.

File: src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs (hunk 8 above)
Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: So after a round of discussion around the software engineering approach to this solution we arrived at the work that was done [f719763](https://github.com/bitwarden/server/pull/6588/commits/f719763a85a489951dfb3a5ad4d20d5cb7138823)

Have a look @quexten and let me know what you think. This approach keeps some attachment to looking at the private key structure which we feel is a more defensive approach than just looking at the SecurityVersion column.

Comment (quexten) [reply src/Core/KeyManagement/Queries/IsV2EncryptionUserQuery.cs]: I guess it's a little confusing, but a V2 is every version after and including 2, not just 2. But it's OK to leave it as is right now, since we don't yet have a direct plan for implementing one of the features that would migrate a V2 user to a security version that is larger than 2.

I'm ok with the solution as-is, thanks for spending the time here!

File: test/Core.Test/KeyManagement/Queries/GetMinimumClientVersionForUserQueryTests.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        Assert.Equal(Core.KeyManagement.Constants.MinimumClientVersionForV2Encryption, version);
+    }
+
+    [Fact]
+    public async Task Run_ReturnsNull_ForV1User()
+    {
+        var sut = new GetMinimumClientVersionForUserQuery();
+        var version = await sut.Run(new User
+        {
+            SecurityVersion = 1,
```
Comment (quexten) [inline test/Core.Test/KeyManagement/Queries/GetMinimumClientVersionForUserQueryTests.cs]: Could we add a test for security version being null?

Comment (Patrick-Pimentel-Bitwarden) [reply test/Core.Test/KeyManagement/Queries/GetMinimumClientVersionForUserQueryTests.cs]: good catch!

Comment (Patrick-Pimentel-Bitwarden) [reply test/Core.Test/KeyManagement/Queries/GetMinimumClientVersionForUserQueryTests.cs]: Added that test and more in [cff2f5d](https://github.com/bitwarden/server/pull/6588/commits/cff2f5df6dd50bf1acbb77e1d971df97108424bb)

File: src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs
Comment (JaredSnider-Bitwarden) [inline src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs:1]: 🎨 : I'm not sure this query is worth the boilerplate. It's not injecting a repository and it doesn't hit the database. It references a `User` object encapsulated method `IsSetupForV2Encryption` and a constant `MinimumClientVersionForV2Encryption` - both of which can be directly accessed in its only use within the `ClientVersionValidator`. Was there something discussed with KM about this needing to be in a re-usable location vs a private method on the `ClientVersionValidator` itself?

File: src/Core/Entities/User.cs [hunk 10]
```diff
@@ -211,6 +212,25 @@ public int GetSecurityVersion()
         return SecurityVersion ?? 1;
     }
 
+    public bool IsSetupForV2Encryption()
```
Comment (JaredSnider-Bitwarden) [inline src/Core/Entities/User.cs]: 🎨 (nit): I find this mostly clear, but there is a small amount of ambiguity in this name. `IsSetupFor` can imply they are setup for a potential migration to V2 encryption, but we actually mean that they are already on V2 encryption if this method returns true. Something like `HasV2Encryption` might clarify this but also a `<summary>` on what the method is intending could work as well.

Comment (JaredSnider-Bitwarden) [reply src/Core/Entities/User.cs]: Aside: the composition of user methods is exactly what I was looking for. Thank you!

File: src/Core/KeyManagement/Utilities/EncryptionParsing.cs
Comment (JaredSnider-Bitwarden) [inline src/Core/KeyManagement/Utilities/EncryptionParsing.cs:1]: 🎨 : Since we have constants for known v1 and v2 key types, can we write a quick unit test suite for this helper?

File: src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs (hunk 1 above)
Comment (JaredSnider-Bitwarden) [reply src/Identity/IdentityServer/RequestValidators/BaseRequestValidator.cs]: This is a great explanation and a good note to leave here so we don't re-order this and open that door in the future.

File: src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs
Comment (JaredSnider-Bitwarden) [reply src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs:1]: I just looked at the ticket and saw this:

> Guidance from Architecture has asked us to consider if we can make this an abstraction/utility such that it can be used in other cases as well where we may have a need in the future to prevent authentication on incompatible versions for different reasons.

```
VersionFeatureCompatibility
{
  MethodToCallToCheckIfFeatureIsEnabledForUser(User u) => boolean { get; set; }
  MinimumVersion => string { get; set; }
}
```

I think a shared utility like this would be cleaner for re-use vs trying to integrate it with the CQS DB approach we have.

File: src/Identity/IdentityServer/RequestValidators/ClientVersionValidator.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+public class ClientVersionValidator(
+    ICurrentContext currentContext,
+    IGetMinimumClientVersionForUserQuery getMinimumClientVersionForUserQuery)
+    : IClientVersionValidator
+{
+    private static readonly string UpgradeMessage = "Please update your app to continue using Bitwarden";
+
+    public async Task<bool> ValidateAsync(User? user, CustomValidatorRequestContext requestContext)
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply src/Identity/IdentityServer/RequestValidators/ClientVersionValidator.cs]: @Patrick-Pimentel-Bitwarden , can you take a look at this feedback? Are requests without a user at this point valid?

File: test/Common/Constants/TestEncryptionConstants.cs
Comment (JaredSnider-Bitwarden) [inline test/Common/Constants/TestEncryptionConstants.cs:1]: Thank you for making these for easy re-use in our tests!

File: test/Common/Constants/TestEncryptionConstants.cs [hunk 12]
```diff
@@ -0,0 +1,19 @@
+﻿namespace Bit.Test.Common.Constants;
+
+public static class TestEncryptionConstants
+{
+    // V1-style encrypted strings (AES-CBC-HMAC formats) accepted by validators
+    public const string V1EncryptedBase64 = "2.QmFzZTY0UGFydA==|QmFzZTY0UGFydA==|QmFzZTY0UGFydA==";
```
Comment (JaredSnider-Bitwarden) [inline test/Common/Constants/TestEncryptionConstants.cs]: ❓ : could we try to come up with a better name here or add to the summary to indicate which types of request fields this is expected to be used in? I had to go look at the tests to understand how it is supposed to be used.

File: test/Identity.IntegrationTest/Login/ClientVersionGateTests.cs
Comment (JaredSnider-Bitwarden) [inline test/Identity.IntegrationTest/Login/ClientVersionGateTests.cs:1]: Great tests and test coverage!

File: src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs
Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs:1]: Ah, yeah see it was originally injecting a repository but then following guidance from KM and conversation with you we just attached it to checking properties on the user. I can take it out of query form.

Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs:1]: Spending some more time thinking about this, I remembered at refinement time I do believe we moved away from the generic VersionFeatureCompatibility as it could also grow the scope of this work more towards platform tooling. Also the VersionFeatureCompatibility is absent from the technical breakdown we wrote. If we are to do a more generic VersionFeatureCompatibility it very may well require repositories in the future and I think it would keep its value better following the CQS DB pattern no?

Happy to sync on this one if you want.

Comment (JaredSnider-Bitwarden) [reply src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs:1]: Ok, that's fair to reduce the scope down to just the validator we have. I still would like to request that we please move the checks into the validator itself as any generalization in the future wouldn't have the same shape as the query anyhow.

File: test/Common/Constants/TestEncryptionConstants.cs (hunk 12 above)
Comment (quexten) [reply test/Common/Constants/TestEncryptionConstants.cs]: I suggest: "AES256_CBC_HMAC_Encstring", and putting the information that this is meant for V1 users into a comment. Technically, even V2 users can have AES256_CBC_HMAC encstrings in their account, for anything that is not directly encrypted by the user keys. Organization ciphers are currently also all AES256_CBC_HMAC.

Comment (Patrick-Pimentel-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: fixed with [2264056](https://github.com/bitwarden/server/pull/6588/commits/226405609e4e5f935eb0c819170c07dc94d7fe9d)

File: src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs
Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Queries/GetMinimumClientVersionForUserQuery.cs:1]: Addressed comments in [2264056](https://github.com/bitwarden/server/pull/6588/commits/226405609e4e5f935eb0c819170c07dc94d7fe9d)

File: src/Core/KeyManagement/Utilities/EncryptionParsing.cs
Comment (Patrick-Pimentel-Bitwarden) [reply src/Core/KeyManagement/Utilities/EncryptionParsing.cs:1]: added in the latest commits [36e7b1c](https://github.com/bitwarden/server/pull/6588/commits/36e7b1c65ec21ef5b4b541fa493c91ef7d7fda3f) and [55bfb71](https://github.com/bitwarden/server/pull/6588/commits/55bfb71bef5a981f15b2153afcd437f886b834c0)

File: src/Identity/IdentityServer/RequestValidators/ClientVersionValidator.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (Patrick-Pimentel-Bitwarden) [reply src/Identity/IdentityServer/RequestValidators/ClientVersionValidator.cs]: Addressed!

File: test/Common/Constants/TestEncryptionConstants.cs [hunk 13]
```diff
@@ -5,6 +5,15 @@ public static class TestEncryptionConstants
     // Intended for use as a V1 encrypted string, accepted by validators
     public const string AES256_CBC_HMAC_Encstring = "2.QmFzZTY0UGFydA==|QmFzZTY0UGFydA==|QmFzZTY0UGFydA==";
 
+    // Simple stubs for other encstring versions used by parsing tests
+    public const string AES256_CBC_B64_Encstring = "0.stub";
+    public const string AES128_CBC_HMACSHA256_B64_Encstring = "1.stub";
```
Comment (quexten) [inline test/Common/Constants/TestEncryptionConstants.cs]: Noting: We don't support this encstring type anymore, it does not exist in the real world and client / sdk support was removed.

File: test/Common/Constants/TestEncryptionConstants.cs [hunk 14]
```diff
@@ -5,6 +5,15 @@ public static class TestEncryptionConstants
...
+    public const string AES256_CBC_HMAC_EmptySuffix = "2.";
+    public const string RSA2048_OAEPSHA256_B64_Encstring = "3.stub";
+    public const string RSA2048_OAEPSHA1_B64_Encstring = "4.stub";
+    public const string RSA2048_OAEPSHA256_HMACSHA256_B64_Encstring = "5.stub";
```
Comment (quexten) [inline test/Common/Constants/TestEncryptionConstants.cs]: Noting: Between the RSA types, only type 4 is known to exist in the real world.

File: test/Common/Constants/TestEncryptionConstants.cs [hunk 15]
```diff
@@ -5,6 +5,15 @@ public static class TestEncryptionConstants
...
+    public const string RSA2048_OAEPSHA1_HMACSHA256_B64_Encstring = "6.stub";
```
Comment (quexten) [inline test/Common/Constants/TestEncryptionConstants.cs]: Could we add a Type 7 string?

File: test/Common/Constants/TestEncryptionConstants.cs (hunk 14 above)
Comment (Patrick-Pimentel-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: Would you like me to leave this comment amongst the testing constants? Or is there another action you would like me to take?

Comment (JaredSnider-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: Let's only add ones that we expect to use

File: test/Common/Constants/TestEncryptionConstants.cs (hunk 13 above)
Comment (JaredSnider-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: Let's remove it then

File: test/Common/Constants/TestEncryptionConstants.cs (hunk 14 above)
Comment (JaredSnider-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: Or we can just keep these as examples

File: test/Common/Constants/TestEncryptionConstants.cs (hunk 13 above)
Comment (JaredSnider-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: Or mark it as deprecated

Comment (quexten) [reply test/Common/Constants/TestEncryptionConstants.cs]: Imo: Mark AES256_CBC_B64_Encstring as deprecated, remove the AES128_CB_HMAC256_B64 encstring.

File: test/Common/Constants/TestEncryptionConstants.cs (hunk 14 above)
Comment (quexten) [reply test/Common/Constants/TestEncryptionConstants.cs]: I think it makes sense to only add the ones we expect to see. We probably eventually want to remove any reference to these other types from both clients and servers, and there is no reason to add new uses of it.

Comment (Patrick-Pimentel-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: fixed with [50298fb](https://github.com/bitwarden/server/pull/6588/commits/50298fbbce4227bbd5c9f90d28fcf4ae30d3edab)

File: test/Common/Constants/TestEncryptionConstants.cs (hunk 15 above)
Comment (Patrick-Pimentel-Bitwarden) [reply test/Common/Constants/TestEncryptionConstants.cs]: added with [79256d7](https://github.com/bitwarden/server/pull/6588/commits/79256d73ee355229ed1e9ecec4885402877449d6)

**Review by JaredSnider-Bitwarden — DISMISSED** [review]

**Review by quexten — DISMISSED** [review]

**Review by prograhamming — APPROVED** [review]

**Review by ike-kottlowski — APPROVED** [review]

**Comment by trmartin4** [comment]
Resolved merge conflicts again.

**Review by ike-kottlowski — APPROVED** [review]

**Review by quexten — APPROVED** [review]

**Review by prograhamming — APPROVED** [review]
